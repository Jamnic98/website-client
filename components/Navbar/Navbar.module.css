.navbar {
  position: sticky;
  height: var(--navbar-height);
  top: 0;
  z-index: 1;
  background-color: var(--c-bg-secondary);
  color: var(--c-text-navbar);
  /* box-shadow: 0 0.2em 0.4em grey; */
}

.navbar > span {
  height: 100%;
  display: flex;
  justify-content: space-between;
  background-color: var(--c-bg-secondary);
  color: inherit;
}

.navlinks {
  display: flex;
  flex-direction: row;
  align-items: center;
  text-align: center;
  width: 30vw;
  max-width: 600em;
  height: 100%;
  font-weight: bold;
  background-color: var(--c-bg-secondary);
}

.navlink:hover {
  background-color: var(--c-link);
}

.logoContainer {
  display: flex;
  align-items: center;
  text-align: center;
  font-weight: bold;
}

.logo {
  font-size: 1.5em;
  color: var(--c-link);
  transition: color 0.3s ease;
  min-width: max-content;
  letter-spacing: 1px;
}

.logo:hover {
  color: var(--c-text-navbar);
}

.menuIcon {
  visibility: hidden;
  font-size: 1.6em;
  cursor: pointer;
  position: absolute;
  left: 0;
  top: 25%;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-left: 0.3em;
  color: var(--c-text-navbar);
}

@media screen and (max-width: 800px) {
  .navlinks {
    flex-direction: column;
    position: absolute;
    display: block;
    width: 100%;
    height: auto;
    top: var(--navbar-height);
    left: -100%;
    background-color: inherit;
    transition: transform 0.3s ease;
  }

  .navlinks.open {
    transform: translateX(100%);
  }

  .menuIcon {
    visibility: unset;
  }

  .logoContainer {
    position: relative;
    margin-left: auto;
    margin-right: auto;
  }
}

@media screen and (max-width: 600px) {
  .navbar {
    font-size: 1em;
  }
}

@media screen and (max-width: 300px) {
  .navbar {
    font-size: 0.8em;
  }
}
