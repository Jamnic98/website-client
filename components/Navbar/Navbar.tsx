import { type FC, useState } from 'react'
import { MobileMenuToggle, Navlink, NavLogo } from '..'
import { type NavlinkData } from '../../types/global'
import classNames from 'classnames'
import styles from './Navbar.module.css'

export interface NavbarProps {
  navlinkObjArr: NavlinkData[]
}

export const Navbar: FC<NavbarProps> = ({ navlinkObjArr }) => {
  const [isOpen, setIsOpen] = useState(false)
  const handleClick = () => setIsOpen(!isOpen)

  return (
    <nav className={styles.navbar} data-testid='navbar' aria-label='navbar'>
      <span className='container'>
        <MobileMenuToggle handleClick={handleClick} isOpen={isOpen} />
        <NavLogo />
        <span
          className={classNames(styles.navlinks, { [styles.open]: isOpen })}
          aria-label='navlinks'
        >
          {navlinkObjArr.map((navlinkObj, index) => (
            <Navlink
              onClick={() => setIsOpen(false)}
              navlinkObj={navlinkObj}
              key={index}
            />
          ))}
        </span>
      </span>
    </nav>
  )
}
