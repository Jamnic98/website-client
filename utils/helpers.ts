const getAverageDurationString = (averageDurationInSeconds: number) =>
  `Average run time: ${secondsToDhms(averageDurationInSeconds)}`;

const getAverageDistanceString = (averageDistanceInMeters: number) => {
  const averageDistanceInKm =
    averageDistanceInMeters === 0 ? 0 : averageDistanceInMeters / 1000;
  return `Average distance: ${averageDistanceInKm.toFixed(2)}km / ${kmToMiles(
    averageDistanceInKm
  ).toFixed(2)}mi`;
};

const getTotalDistanceString = (totalDistanceInMeters: number): string => {
  const totalDistanceInKm = totalDistanceInMeters / 1000;
  return `Total distance: ${totalDistanceInKm.toFixed(2)}km / ${kmToMiles(
    totalDistanceInKm
  ).toFixed(2)}mi`;
};

const getFurthestDistanceString = (
  furthestDistanceInMeters: number
): string => {
  const distanceInKm =
    furthestDistanceInMeters === 0 ? 0 : furthestDistanceInMeters / 1000;
  return `Furthest distance: ${distanceInKm.toFixed(2)}km / ${kmToMiles(
    distanceInKm
  ).toFixed(2)}mi`;
};

const getTotalDurationString = (totalDuration: number) =>
  `Total duration: ${secondsToDhms(totalDuration)}`;

const kmToMiles = (distanceInKm: number) => 0.621371 * distanceInKm;

const secondsToDhms = (seconds: number): string => {
  if (seconds === 0) {
    return '-';
  } else {
    const d = Math.floor(seconds / (3600 * 24));
    const h = Math.floor((seconds % (3600 * 24)) / 3600);
    const m = Math.floor((seconds % 3600) / 60);
    const s = Math.floor(seconds % 60);

    const dDisplay = d > 0 ? d + 'D' : '';
    const hDisplay = h > 0 ? h + 'h' : '';
    const mDisplay = m > 0 ? m + 'm' : '';
    const sDisplay = s > 0 ? s + 's' : '';
    return `${dDisplay} ${hDisplay} ${mDisplay} ${sDisplay}`;
  }
};

const reduceSumFunc = (prev: number, cur: number) => prev + cur;

const getRandomElement = (array: any[]) =>
  array[Math.floor(Math.random() * array.length)];

export {
  reduceSumFunc,
  getAverageDistanceString,
  getAverageDurationString,
  getTotalDistanceString,
  getFurthestDistanceString,
  getTotalDurationString,
  kmToMiles,
  secondsToDhms,
  getRandomElement,
};
