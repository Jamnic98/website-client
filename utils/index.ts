export * from "./constants";
export * from "./helpers";
export * from "./runs";
export * from "./useWidth";
