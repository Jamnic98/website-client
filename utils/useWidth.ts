import { useEffect, useState } from 'react';

type WindowDimensions = number | undefined;

const useWidth = (): WindowDimensions => {
  const [windowDimensions, setWindowDimensions] =
    useState<WindowDimensions>(undefined);

  useEffect(() => {
    const handleResize = (): void => {
      setWindowDimensions(window.innerWidth);
    };

    handleResize();
    window.addEventListener('resize', handleResize);
    return (): void => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
};

export default useWidth;
