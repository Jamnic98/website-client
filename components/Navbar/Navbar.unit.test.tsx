import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'

import { Navbar } from './Navbar'
import { navlinkObjArr } from '../../utils'

// TODO: remove
/* const testNavlinkObjArr = [
  {
    label: 'logs',
    url: '/logs/running',
    children: [
      { label: 'running', url: '/logs/running', children: [] }
      // { label: 'weight lifting', url: '/logs/weight-lifting', children: [] },
      // { label: 'habit tracker', url: '/logs/habit-tracker', children: [] },
      // { label: 'reading', url: '/logs/reading', children: [] }
    ]
  },
  { label: 'projects', url: '/projects', children: [] },
  { label: 'about ', url: '/about', children: [] }
]

const getNavbar = (props?: NavbarProps) => (
  <Navbar navlinkObjArr={testNavlinkObjArr} />
) */

describe('Navbar', () => {
	it('renders a Navbar', () => {
		render(<Navbar navlinkObjArr={navlinkObjArr} />)
		const navbar = screen.getByTestId('application-navbar')
		expect(navbar).toBeInTheDocument()
	})

	it('renders Navbar unchanged', () =>
		expect(
			renderer.create(<Navbar navlinkObjArr={navlinkObjArr} />).toJSON()
		).toMatchSnapshot())
})
