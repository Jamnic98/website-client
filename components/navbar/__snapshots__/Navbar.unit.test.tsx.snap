// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders navbar unchanged 1`] = `
<nav
  aria-label="navbar"
  className="navbar"
  data-testid="navbar"
>
  <span
    className="container"
  >
    <span
      aria-label="mobile menu toggle button"
      className="menuIcon"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </span>
    <a
      className="navLogo"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Jamie Stimpson
    </a>
    <span
      aria-label="navlinks"
      className="navlinks"
    >
      <div
        className="dropdownNavlink"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <a
          className="navlink"
          href="/logs/running"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          logs
          <span
            className="caret up"
          >
            ^
          </span>
        </a>
        <div
          className="dropdownContainer"
          data-testid="dropdownContainer"
        >
          <div
            className="dropdown closed"
            data-testid="dropdown"
          >
            <div
              className="dropdownItem"
            >
              <a
                className="navlink"
                href="/logs/running"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                running
              </a>
            </div>
          </div>
        </div>
      </div>
      <a
        className="navlink"
        href="/projects"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        projects
      </a>
      <a
        className="navlink"
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        about 
      </a>
    </span>
  </span>
</nav>
`;
